src/coreclr/src/System.Private.CoreLib/src/System/GC.cs

```diff
@@ -653,29 +661,63 @@ internal static void UnregisterMemoryLoadChangeNotification(Action notification)
+        /// pinned: true means you want to pin this object that you are allocating
+        /// otherwise it's not pinned.
+        /// pinned requires that T does not contain object references.
         /// </summary>
         [MethodImpl(MethodImplOptions.AggressiveInlining)] // forced to ensure no perf drop for small memory buffers (hot path)
-        internal static T[] AllocateUninitializedArray<T>(int length)
+        public static T[] AllocateUninitializedArray<T>(int length, bool pinned = false)
         {
-            if (RuntimeHelpers.IsReferenceOrContainsReferences<T>())
+            if (!pinned)
             {
-                return new T[length];
-            }
+                if (RuntimeHelpers.IsReferenceOrContainsReferences<T>())
+                {
+                    return new T[length];
+                }
 
-            // for debug builds we always want to call AllocateNewArray to detect AllocateNewArray bugs
+                // for debug builds we always want to call AllocateNewArray to detect AllocateNewArray bugs
 #if !DEBUG
-            // small arrays are allocated using `new[]` as that is generally faster.
-            if (length < 2048 / Unsafe.SizeOf<T>())
+                // small arrays are allocated using `new[]` as that is generally faster.
+                if (length < 2048 / Unsafe.SizeOf<T>())
+                {
+                    return new T[length];
+                }
+#endif
+            }
+            else if (RuntimeHelpers.IsReferenceOrContainsReferences<T>())
             {
-                return new T[length];
+                ThrowHelper.ThrowInvalidTypeWithPointersNotSupported(typeof(T));
             }
-#endif
+
             // kept outside of the small arrays hot path to have inlining without big size growth
-            return AllocateNewUninitializedArray(length);
+            return AllocateNewUninitializedArray(length, pinned);
 
             // remove the local function when https://github.com/dotnet/coreclr/issues/5329 is implemented
-            static T[] AllocateNewUninitializedArray(int length)
-                => Unsafe.As<T[]>(AllocateNewArray(typeof(T[]).TypeHandle.Value, length, zeroingOptional: true));
+            static T[] AllocateNewUninitializedArray(int length, bool pinned)
+            {
+                GC_ALLOC_FLAGS flags = GC_ALLOC_FLAGS.GC_ALLOC_ZEROING_OPTIONAL;
+                if (pinned)
+                    flags |= GC_ALLOC_FLAGS.GC_ALLOC_PINNED_OBJECT_HEAP;
+
+                return Unsafe.As<T[]>(AllocateNewArray(typeof(T[]).TypeHandle.Value, length, flags));
+            }
+        }
+
+        /// <summary>
+        /// pinned: true means you want to pin this object that you are allocating
+        /// otherwise it's not pinned.
+        /// pinned requires that T does not contain object references.
+        /// </summary>
+        public static T[] AllocateArray<T>(int length, bool pinned = false)
```

Comment:
Wouldn't be better to have this as `AllocatePinnedArray<T>` only? What is the difference between `new object[10]` and `AllocateArray<object>(10)` ?

Reply:
There are plans to add more overloads to this API, possibly to control target generation or alignment.

Indeed, both times when this was reviewed, it was noticed that the trivial case is the same as 'new t[]`, but it seems to be ok and happens with other APIs that have optional parameters.

Reply:
It seems odd to me that every time someone uses this API one has to type the explicit parameter to make the code readable.

Reply:
I expect that this API is going to be used very rarely. The more useful API build on top this is going to be a pool for the pinned arrays. The API for that was not designed yet. @VSadov Is it in your plans - what do you expect the pool API to be?

Reply:
Arguments for going it this way were convincing. When there are more arguments, a new overload will be introduced with 2 optional parameters, while this one becomes not optional and so on...

Reply:
For the pool - I think from the usability point, it might be preferable to just add `GetPinned` to the current pool API. 
Alternatively we could have a separate pool for pinned arrays, but then it puts a burden on the user to remember where to return.